Prevent failed KafkaConsumer creation from blocking overlord startup

jon-wei opened the pull request:
Taking a stab at fixing #6114 

This PR changes the initialization behavior for KafkaSupervisor so that it does not block the lifecycle if supervisor init fails. The supervisor will retry the initialization indefinitely.

Marking WIP as it needs more testing, but opening it in case anyone has comments about the approach.

b-slim commented:
How about changing the workflow and switch to a consumerProvider that can be used at runtime (when needed), the provider can be supplier with memoization or something similar if the consumer is to be reused, and the task that needs it will fail if the provider times-out.

jon-wei commented:
@b-slim I initially thought about lazily creating the consumer, but decided that the supervisor couldn't really do anything useful if it couldn't get a consumer, so I think it makes sense to have the supervisor not finish initialization, so it's as if the supervisor isn't there (seems like a simpler failure mode than "running but with invalid consumer")

jihoonson commented on extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/indexing/kafka/supervisor/KafkaSupervisor.java:
```diff
@@ -416,61 +421,87 @@ public void start()
+      // Try normal initialization first, if that fails then schedule periodic initialization retries
+      tryInit();
+      if (!started) {
+        initializationExec.scheduleAtFixedRate(
```
How about using `RetryUtils.retry()`? It has a backoff mechanism which I think makes sense to apply here.

jon-wei replied:
Since there are checks for submitting an invalid supervisor spec (e.g., if nothing in "bootstrap.servers" is resolvable), the error condition this is addressing seems to be situations where some external state changes and the supervisor spec no longer works. For the unresolvable bootstrap.servers issue as an example, maybe its a transient DNS error, so I feel like it's better to retry indefinitely without increasing backoff times (with retries tied to the configured supervisor run period instead).

RetryUtils.retry() is also a blocking call, which I want to avoid here in the lifecycle startup.

jihoonson replied:
Hmm, I think backoff is more appropriate for transient errors. It starts with  retrying in a very short interval at first, but the retry interval grows as the failure continues. This makes sense because it can recover quickly with short retry periods if the error is transient. If the error is permanent, the retry period would grow in the end and it won't use excessive resources. Regarding too large backoff time, `RetryUtils.retry()` has the `MAX_SLEEP_MILLIS` which I think it makes sense to make it configurable. Regarding blocking call, please check my below comment. What do you think?

jihoonson replied:
Also, does it make sense to add a sort of max limit for retrying?

jon-wei replied:
Hm, I suppose the user can resubmit the supervisor in the situation where the backoff gets extremely large, I'll change to have limited retries with backoff

jihoonson commented on extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/indexing/kafka/supervisor/KafkaSupervisor.java:
```diff
@@ -260,6 +261,7 @@ public String toString()
 
   private volatile DateTime firstRunTime;
   private volatile KafkaConsumer consumer;
+  private volatile boolean lifecycleStarted = false;
```
How about adding a new `InitNotice`? It can be processed by the existing `exec`, so we don't have to add a new `initializationExec`.

jon-wei replied:
I would need to start the notice handler thread to use InitNotice, which I was trying to avoid doing (the idea of the patch was to keep the supervisor in a state similar to what it would have been in if it were stuck in the current lifecycle start, while allowing the rest of the lifecycle to start).

jihoonson replied:
> the idea of the patch was to keep the supervisor in a state similar to what it would have been in if it were stuck in the current lifecycle start, while allowing the rest of the lifecycle to start

Got it. Makes sense. 

The reason I commented about `InitNotice` was, I'm not sure `initializationExec` is needed. Maybe we can use `scheduledExec` instead and cancel the scheduledFuture once `tryInit` succeeds? Or we can submit a runnable task to `exec` which calls `RetryUtils.retry()`.

jon-wei replied:
Sure, I'll change to submit a runnable to exec for the init retries

jon-wei commented:
Addressed comments.

I also found an issue with `RealtimeIndexTask.makeRandomId()` introduced in #6226, the random portion of the task ID was shortened from 8 -> 1, this PR fixes that

jihoonson commented on indexing-service/src/main/java/org/apache/druid/indexing/common/task/RealtimeIndexTask.java:
```diff
@@ -89,7 +89,7 @@
   private static final int TASK_ID_BITS_PER_SYMBOL = 4;
   private static final int TASK_ID_SYMBOL_MASK = (1 << TASK_ID_BITS_PER_SYMBOL) - 1;
-  private static final int TASK_ID_LENGTH = Integer.BYTES / TASK_ID_BITS_PER_SYMBOL;
+  private static final int TASK_ID_LENGTH = (Integer.BYTES * 8) / TASK_ID_BITS_PER_SYMBOL;
```
nit: `Integer.SIZE` returns 32.

jon-wei replied:
changed to Integer.SIZE
